sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/encoders/ExpressionEncoder.scala

```diff
@@ -47,16 +47,26 @@ object ExpressionEncoder {
     // We convert the not-serializable TypeTag into StructType and ClassTag.
     val mirror = typeTag[T].mirror
     val tpe = typeTag[T].tpe
+
+    if (ScalaReflection.optionOfProductType(tpe)) {
+      throw new UnsupportedOperationException(
+        "Cannot create encoder for Option of Product type, because Product type is represented " +
```

Comment:
and now that i think about it more, i also think Dataset[Option[(Int, Int)]] should be valid too if possible. 

it should not be represented by a top level Row object, so the schema should be
```StructType(StructField("_1", StructType(StructField("_1", IntegerType, false), StructField("_2", IntegerType, false)), true))```

basically it should be treated as flat. we do this trick for flat types all the time (nest them inside a row with one element), so why not for an Option[X <: Product]?